### PR TITLE
Refactor assembler/disassembler to share internal code representation.

### PR DESCRIPTION
**Pull request by ribrdb**
#29

### CODE REVIEW

File: ir/instruction.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
+export interface LiteralValue {
```
Comment (quisar) [inline ir/instruction.ts:1]: Is it important to use interfaces? It makes the isX() method a little bit brittle. If these where classes we could have a Symbol for each type, and a type variable with the correct symbol.

File: ir/instruction.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+interface Label {
+  label: string;
+}
+
+export function parseLabel(x: string): Label {
+  if (x.startsWith(":")) {
+    return { label: x.substring(1) };
+  } else if (x.endsWith(":")) {
```
Comment (quisar) [inline ir/instruction.ts:24]: Where are we using labels ending with ':' ?

File: decompile/disasm.ts [hunk 3]
```diff
@@ -77,121 +106,96 @@ export class Disassembler {
...
-    return labels;
+    this.program.code[this.program.code.length - 1].next ??= { stop: true };
   }
 
   #emitInstr(
-    inst: RawInstruction,
+    raw: RawInstruction,
     ip: number,
-    labels: Map<number, string>,
+    nodeOffset,
```
Comment (quisar) [inline decompile/disasm.ts:127]: Don't we want to type nodeOffset?

File: ir/instruction.ts (hunk 2 above)
Comment (ribrdb) [reply ir/instruction.ts:24]: I meant this to work for a label for a line in assembly and a label ref as an argument. But i never ended up using this function

File: ir/instruction.ts (hunk 1 above)
Comment (ribrdb) [reply ir/instruction.ts]: I couldn't get it to work using Symbol, but I think this is the same idea.

**Review by quisar — APPROVED** [review]

**Pull request merged**